src/flamenco/repair/fd_recorder.c

```diff
@@ -0,0 +1,732 @@
+      }
+
+    }
+}
+
+/* Helper structure for sorting peers by RTT */
+struct peer_rtt_pair {
+  fd_pubkey_t * key;
+  double        rtt;
+};
+
+/* Comparison function for qsort - sorts by RTT ascending (quickest first) */
+static int
+peer_rtt_compare( const void * a, const void * b ) {
+  const struct peer_rtt_pair * pair_a = (const struct peer_rtt_pair *)a;
+  const struct peer_rtt_pair * pair_b = (const struct peer_rtt_pair *)b;
+
+  if( pair_a->rtt < pair_b->rtt ) return -1;
+  if( pair_a->rtt > pair_b->rtt ) return 1;
+  return 0;
+}
+
+/* Helper function to reshuffle peers into categories based on count.
+   First 500 quickest -> high priority
+   Next 500 quickest -> medium priority
+   Remaining -> low priority
+   Zero hit rate peers stay separate */
+void
+fd_recorder_reshuffle_peers_cnt( fd_recorder_t * recorder ) {
+  recorder->high_priority_cnt = 0;
+  recorder->medium_priority_cnt = 0;
+  recorder->low_priority_cnt = 0;
+  recorder->zero_hr_cnt = 0;
+
+  recorder->high_priority_idx = 0;
+  recorder->medium_priority_idx = 0;
+  recorder->low_priority_idx = 0;
+  recorder->zero_hr_idx = 0;
+
+  fd_recorder_peer_map_t * peer_map = fd_recorder_peer_map( recorder );
+  fd_recorder_peer_t * peer_pool = fd_recorder_peer_pool( recorder );
+
+  /* First pass: count total peers and separate zero HR peers */
+  ulong non_zero_hr_peers = 0;
+
+  for( fd_recorder_peer_map_iter_t iter = fd_recorder_peer_map_iter_init( peer_map, peer_pool );
+       !fd_recorder_peer_map_iter_done( iter, peer_map, peer_pool );
+       iter = fd_recorder_peer_map_iter_next( iter, peer_map, peer_pool ) ) {
+
+    fd_recorder_peer_t * peer = fd_recorder_peer_map_iter_ele( iter, peer_map, peer_pool );
+
+    /* Separate zero hit rate peers immediately */
+    if( peer->ewma_hr <= 0.0 ) recorder->zero_hr_peers[recorder->zero_hr_cnt++] = &peer->key;
+    else                       non_zero_hr_peers++;
+  }
+
+  /* If no non-zero HR peers, we're done */
+  if( FD_UNLIKELY( non_zero_hr_peers == 0 ) ) return;
+
+  /* Allocate temporary array for sorting non-zero HR peers */
+  struct peer_rtt_pair sorted_peers[FD_MAX_PEERS];
+  ulong sorted_count = 0;
+
+  /* Second pass: collect non-zero HR peers for sorting */
+  for( fd_recorder_peer_map_iter_t iter = fd_recorder_peer_map_iter_init( peer_map, peer_pool );
+       !fd_recorder_peer_map_iter_done( iter, peer_map, peer_pool );
+       iter = fd_recorder_peer_map_iter_next( iter, peer_map, peer_pool ) ) {
+
+    fd_recorder_peer_t * peer = fd_recorder_peer_map_iter_ele( iter, peer_map, peer_pool );
+
+    /* Skip zero HR peers (already handled) */
+    if( FD_UNLIKELY( peer->ewma_hr <= 0.0 ) ) continue;
+
+    /* Add to sorting array */
+    sorted_peers[sorted_count].key = &peer->key;
+    sorted_peers[sorted_count].rtt = peer->ewma_rtt;
+    sorted_count++;
+  }
+
+  /* Sort peers by RTT (quickest first) */
+  qsort( sorted_peers, sorted_count, sizeof(struct peer_rtt_pair), peer_rtt_compare );
```

Comment:
fd_sort not qsort. also instead of this periodic reshuffle op, consider a way to incrementally update with the right data structure?